Mathlib/NumberTheory/LSeries/SumCoeff.lean

```diff
@@ -153,3 +160,222 @@ theorem LSeries_eq_mul_integral_of_nonneg (f : ℕ → ℝ) {r : ℝ} (hr : 0 
   LSeries_eq_mul_integral' _ hr hs <| hO.congr_left fun _ ↦ by simp [_root_.abs_of_nonneg (hf _)]
 
 end integralrepresentation
+
+noncomputable section residue
+
+variable {f : ℕ → ℂ} {l : ℂ} (hlim : Tendsto (fun n : ℕ ↦ (∑ k ∈ Icc 1 n, f k) / n) atTop (𝓝 l))
+
+section lemmas
+
+include hlim in
+private theorem lemma₁ {s : ℝ} (hs : 1 < s) :
+    IntegrableOn (fun t : ℝ ↦ (∑ k ∈ Icc 1 ⌊t⌋₊, f k) * (t : ℂ) ^ (-(s : ℂ) - 1)) (Set.Ici 1) := by
+  have h₁ : LocallyIntegrableOn (fun t : ℝ ↦ (∑ k ∈ Icc 1 ⌊t⌋₊, f k) * (t : ℂ) ^ (-(s : ℂ) - 1))
+        (Set.Ici 1) := by
+    simp_rw [mul_comm]
+    refine locallyIntegrableOn_mul_sum_Icc f zero_le_one ?_
+    refine ContinuousOn.locallyIntegrableOn (fun t ht ↦ ?_) measurableSet_Ici
+    exact (continuousAt_ofReal_cpow_const _ _ <|
+      Or.inr (zero_lt_one.trans_le ht).ne').continuousWithinAt
+  have h₂ : (fun t : ℝ ↦ ∑ k ∈ Icc 1 ⌊t⌋₊, f k) =O[atTop] fun t ↦ t ^ (1 : ℝ) := by
+    simp_rw [Real.rpow_one]
+    refine IsBigO.trans_isEquivalent ?_ isEquivalent_nat_floor
+    suffices Tendsto (fun n ↦ (∑ k ∈ Icc 1 n, f k) / ((n : ℝ) ^ (1 : ℝ) : ℝ)) atTop (𝓝 l) by
+      simpa using (isBigO_atTop_natCast_rpow_of_tendsto_div_rpow this).comp_tendsto
+        tendsto_nat_floor_atTop
+    simpa using hlim
+  refine h₁.integrableOn_of_isBigO_atTop (g := fun t ↦ t ^(-s)) ?_ ?_
+  · refine IsBigO.mul_atTop_rpow_of_isBigO_rpow 1 (-s - 1) _ h₂ ?_ (by linarith)
+    exact (norm_ofReal_cpow_eventually_eq_atTop _).isBigO.of_norm_left
+  · rwa [integrableAtFilter_rpow_atTop_iff, neg_lt_neg_iff]
+
+private theorem lemma₂ {s T ε : ℝ} {S : ℝ → ℂ} (hs : 1 < s) (hT : 0 < T)
+    (hS₁ : LocallyIntegrableOn (fun t ↦ S t) (Set.Ici 1)) (hS₂ : ∀ t ≥ T, ‖S t‖ ≤ ε * t) :
+    IntegrableOn (fun t : ℝ ↦ ‖S t‖ * (t ^ (-s - 1))) (Set.Ici 1) := by
+  have h : LocallyIntegrableOn (fun t : ℝ ↦ ‖S t‖ * (t ^ (-s - 1))) (Set.Ici 1) := by
+    refine hS₁.norm.mul_continuousOn ?_ isLocallyClosed_Ici
+    exact fun t ht ↦ (Real.continuousAt_rpow_const _ _
+      <| Or.inl (zero_lt_one.trans_le ht).ne').continuousWithinAt
+  refine h.integrableOn_of_isBigO_atTop (g := fun t ↦ t ^(-s)) (isBigO_iff.mpr ⟨ε, ?_⟩) ?_
+  · filter_upwards [eventually_ge_atTop T] with t ht
+    have ht' : 0 < t := hT.trans_le ht
+    rw [norm_mul, norm_norm, show t ^ (-s) = t * t ^ (-s - 1) by
+        rw [Real.rpow_sub ht', Real.rpow_one, mul_div_cancel₀ _ ht'.ne'],
+      norm_mul, Real.norm_of_nonneg ht'.le, ← mul_assoc]
+    exact mul_le_mul_of_nonneg_right (hS₂ t ht) (norm_nonneg _)
```

Comment:
If you do `filter_upwards [eventually_ge_atTop T, eventually_ge_atTop 0] with t ht ht'` you don't need to make `ht'` separately.